mne/epochs.py

```diff
@@ -2479,8 +2480,183 @@ def _ensure_list(x):
+    for idx in range(len(events)):
+        # get the beginning and end sample of the epoch window
+        onset_samp = np.rint(events[idx, 0] - times[0] * raw_sfreq).astype(int)
+        offset_samp = np.rint(events[idx, 0] +
+                              times[-1] * raw_sfreq).astype(int)
+
+        epoch_wins[idx, :] = [onset_samp, offset_samp]
+
+    # get all the indices of onset within Epoch windows
+    onset_indices = np.argwhere((epoch_wins[:, 0] < annot_onset_samp) &
+                                (epoch_wins[:, 1] > annot_onset_samp))
+
+    # get all the indices of offset within Epoch windows
+    annot_offset_samp = annot_onset_samp + annot_duration_samp
+    offset_indices = np.argwhere((epoch_wins[:, 0] < annot_offset_samp) &
+                                 (epoch_wins[:, 1] > annot_offset_samp))
+
+    # get the set of all indices
+    epoch_indices = np.unique(np.concatenate((onset_indices, offset_indices)))
+    return epoch_indices
+
+
+class AnnotationsMixin():
+    """Mixin class for Annotations in Epochs."""
+
+    @property
+    def annotations(self):  # noqa: D102
+        return self._annotations
+
+    @verbose
+    def _set_annotations(self, annotations, on_missing='raise'):
+        """Setter for Epoch annotations from Raw.
+
+        This private function simply copies over Raw annotations, and
+        does not handle offsetting the times based on first_samp
+        or measurement dates, since that is expected to occur in
+        Raw.set_annotations().
+
+        Parameters
+        ----------
+        annotations : instance of mne.Annotations | None
+            Annotations to set. If None, the annotations is defined
+            but empty.
+        %(on_missing_ch_names)s
+
+        Returns
+        -------
+        self : instance of Raw
+            The raw object with annotations.
+        """
+        if annotations is None:
+            self._annotations = None
+        else:
+            _validate_type(annotations, Annotations, 'annotations')
+            new_annotations = annotations.copy()
+            new_annotations._prune_ch_names(self.info, on_missing)
+            self._annotations = new_annotations
+
+        return self
+
+    def get_epoch_annotations(self):
+        """Return a list of raw annotations per epoch.
+
+        Returns
+        -------
+        epoch_annots : list
+            A list (with length equal to number of epochs) of annotations per
+            epoch. Each annotation is stored as (onset, duration, description),
+            where the onset is now relative to the epoch, rather then in
+            continuous time.
+        """
+        events = self.events
+
+        # create a list of annotations for each epoch
+        epoch_annot_list = [[] for _ in range(len(events))]
+
+        # check if annotations exist
+        if self.annotations is None:
+            return epoch_annot_list
+
+        for annot in self.annotations:
```

Comment:
I still think you can eliminate the for-loop.

```py
epoch_tzeros = self.events[:, 0] / self._raw_sfreq  # no need to account for first_samp here
epoch_starts, epoch_stops = np.atleast_2d(epoch_tzeros) + np.atleast_2d(self.times[[0, -1]]).T
annot_starts = self._annotations.onset  # ... because first_samp isn't accounted for here either
annot_stops = annot_starts + self._annotations.duration

annot_spans_epoch_start = np.logical_and(
    np.atleast_2d(epoch_starts) > np.atleast_2d(annot_starts).T,
    np.atleast_2d(epoch_starts) < np.atleast_2d(annot_stops).T)

annot_spans_epoch_end = np.logical_and(
    np.atleast_2d(epoch_stops) > np.atleast_2d(annot_starts).T,
    np.atleast_2d(epoch_stops) < np.atleast_2d(annot_stops).T)

annot_fully_within_epoch = np.logical_and(
    np.atleast_2d(epoch_starts) < np.atleast_2d(annot_starts).T,
    np.atleast_2d(epoch_stops) > np.atleast_2d(annot_stops).T)

all_cases = (annot_spans_epoch_start + annot_spans_epoch_end +
             annot_fully_within_epoch)

for annot_ix, epo_ix in zip(np.where(all_cases)):
    this_annot = self._annotations[annot_ix]
    this_tzero = epoch_tzeros[epo_ix]
    annot = (this_annot['onset'] - this_tzero,
             this_annot['duration'],
             this_annot['description'])
    epoch_annot_list[epo_ix].append(annot)
```

To me this seems more straightforward (e.g. no need to deal with first_samp) and the variable names make clear what the numpy functions are doing.  But maybe I'm oversimplifying  / missing something?

Reply:
Ah I see yeah. I tested this out and this is way more concise. I added this in with a few minor syntax fixes and this actually fixed the issue you raised in your comment.